Initial gcc toolchain configuration for the lowRISC compiler.

cfrantz opened the pull request:
Configure lowrisc_toolchain_rv32imc for use by the OpenTitan project.

cfrantz committed:
Initial toolchain definition for riscv32

Configure lowrisc_toolchain_rv32imc for use by the OpenTitan project.

Signed-off-by: Chris Frantz <cfrantz@google.com>

cfrantz committed:
WIP Fixes for building OpenTitan

1. Deliver per-device flags to the GccEmbeddedFeatures function.
2. Turn off some per-feature compiler/linker flags.  Need to find a
   better way to do this.

Signed-off-by: Chris Frantz <cfrantz@google.com>

cfrantz commented:
Things to discuss:
1) What is the proper name in `constraints/cpu/BUILD`: riscv32 or rv32imc?  Based on the [bazelbuild/platforms](https://github.com/bazelbuild/platforms/blob/main/cpu/BUILD#L123) repo, I feel like `riscv32` is the name of the cpu and `rv32imc` is a more specific architecture-and-features descriptor.

2) Need to do better on compiler flag selections in `toolchains/features/common/impl/gcc.bzl` and `toolchains/features/embedded/impl/gcc.bzl`.  I think these flag selections should be data driven and the flags themselves should be moved into the platform or device configurations: `platforms/BUILD` and/or `devices/riscv32/riscv32.bzl`.  This is a bit of a bigger structural change and will probably require some consultation with the upstream maintainers.

mundaym commented on toolchains/features/common/impl/gcc.bzl:
```diff
@@ -49,31 +49,56 @@ def GccIncludeFeature(include_paths):
-    )
+    if architecture == 'rv32imc':
+        # TODO(cfrantz): why do I have to do this?  Seems the lowRISC gcc
+        # doesn't support `-mfpu`, `-mfloat-abi` or `-mlittle-endian`.
```
`-mlittle-endian` should work (at least it should if we update gcc). AFAICT the other flags are arm specific and aren't defined for other targets. The flag definitions should probably be moved into arch-specific code.

mundaym commented on constraints/cpu/BUILD:
```diff
@@ -15,3 +15,12 @@ constraint_value(
+constraint_value(
+    # Based on the current riscv64 consrtaint in platforms, I wonder
+    # if this should be `riscv32` or `rv32imc`.
+    name = "rv32imc",
```
I'm leaning towards keeping `rv32imc`. It looks like the generic `arm` target is slated for removal and there are a bunch of more specific 32-bit arm targets. `riscv64` is probably geared towards the non-embedded use case.

cfrantz replied:
Based on https://github.com/bazelbuild/platforms/pull/30, `riscv32` is the appropriate high-level cpu descriptor.  In fact, the change has already been applied by Google's copybara automation to Google internal copy of the `platforms` repo.

cfrantz committed:
Rename the cpu architecture as `riscv32`.

This naming is consistent with bazel's `platforms` which names
high-level cpu architectures but not architecture details or cpu
features.

Signed-off-by: Chris Frantz <cfrantz@google.com>

cfrantz commented:
To be consistent with bazelbuild/platforms#30, I've renamed the cpu to `riscv32`.

cfrantz commented:
I think we should commit this change as-is to move the bazel work foward.

I have opened https://github.com/silvergasp/bazel-embedded/issues/37 with upstream to try to get feedback on the proper approach to integrating compiler/feature flags into a "user-serviceable" part of the bazel-embedded repo.  Until we have that understanding, its fine to live with the `if architecture == 'riscv32'` hacks in `toolchains/features/{common, embedded}/impl/gcc.bzl`.

mundaym commented:
> I think we should commit this change as-is to move the bazel work foward.

SGTM. Feel free to commit this.

mcy commented on toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl:
```diff
@@ -0,0 +1,69 @@
+        "defs.bzl",
+        """
+SYSTEM_INCLUDE_COMMAND_LINE = {}
+SYSTEM_INCLUDE_PATHS= {}
```
Space after the = sign? =P

cfrantz replied:
Fixed.  These two formatting errors also exist in the file I copied this from (`gcc_arm_none_repository.bzl`)

mcy commented on toolchains/compilers/lowrisc_toolchain_rv32imc/lowrisc_toolchain_rv32imc_repository.bzl:
```diff
@@ -0,0 +1,69 @@
+        """
+filegroup(
+    name = "all",
+    srcs = glob(["**/*"],exclude=["**/*.html","**/*.pdf"]),
```
Missing space after the , too

cfrantz replied:
Fixed.

mcy reviewed (commented):
Just a couple of format nits. I'm going to assume Michael's LGTM is good enough. =)
